pkg/datapath/loader/netkit.go

```diff
@@ -0,0 +1,187 @@
+//
+// attach is either ebpf.AttachNetkitPrimary or ebpf.AttachNetkitPeer and
+// will attach the program to the xmit of either the primary or peer device.
+func attachNetkit(device netlink.Link, prog *ebpf.Program, progName, bpffsDir string, attach ebpf.AttachType) error {
+	if err := bpf.MkdirBPF(bpffsDir); err != nil {
+		return fmt.Errorf("creating bpffs link dir for netkit attachment to device %s: %w", device.Attrs().Name, err)
+	}
+
+	l, err := link.AttachNetkit(link.NetkitOptions{
+		Program:   prog,
+		Attach:    attach,
+		Interface: device.Attrs().Index,
+		Anchor:    link.Tail(),
+	})
+	if err != nil {
+		return fmt.Errorf("attaching netkit: %w", err)
+	}
+
+	defer func() {
+		// The program was successfully attached using netkit. Closing
+		// a link does not detach the program if the link is pinned.
+		if err := l.Close(); err != nil {
+			log.Warnf("Failed to close netkit link for program %s", progName)
+		}
+	}()
+
+	pin := filepath.Join(bpffsDir, progName)
+	if err := l.Pin(pin); err != nil {
+		return fmt.Errorf("pinning link at %s for program %s : %w", pin, progName, err)
+	}
+
+	log.Infof("Program %s attached to device %s using netkit", progName, device.Attrs().Name)
+
+	return nil
+}
+
+// updateNetkit attempts to update an existing netkit link called progName
+// in bpffsDir. If the link is defunct, the pin is removed.
+//
+// Returns nil if the update was successful. Returns an error wrapping
+// [os.ErrNotExist] if the link is defunct or missing.
+func updateNetkit(prog *ebpf.Program, progName, bpffsDir string) error {
+	// Attempt to open and update an existing link.
+	pin := filepath.Join(bpffsDir, progName)
+	err := bpf.UpdateLink(pin, prog)
+	switch {
+	// Link exists, but is defunct, and needs to be recreated. The program
+	// no longer gets triggered at this point and the link needs to be removed
+	// to proceed.
+	case errors.Is(err, unix.ENOLINK):
+		if err := os.Remove(pin); err != nil {
+			return fmt.Errorf("unpinning defunct link %s: %w", pin, err)
+		}
+
+		log.Infof("Unpinned defunct link %s for program %s", pin, progName)
+
+		// Wrap in os.ErrNotExist so the caller needs to look for one error.
+		return fmt.Errorf("unpinned defunct link: %w", os.ErrNotExist)
+
+	// No existing link found, continue trying to create one.
+	case errors.Is(err, os.ErrNotExist):
+		log.Debugf("No existing link found at %s for program %s", pin, progName)
+		return err
+
+	case err != nil:
+		return fmt.Errorf("updating link %s for program %s: %w", pin, progName, err)
+	}
+
+	log.Infof("Updated link %s for program %s", pin, progName)
+	return nil
+}
+
+// detachNetkit attempts to open the link at progName in bpffsDir
+// and unpins it. Only returns unrecoverable errors. Returns nil if
+// the link doesn't exist or if removal was successful.
+func detachNetkit(bpffsDir, progName string) error {
+	pin := filepath.Join(bpffsDir, progName)
+	err := bpf.UnpinLink(pin)
+	if err == nil {
+		log.Infof("Removed netkit link at %s", pin)
```

Comment:
Maybe a bit noisy? I guess you will get an Info message for every pod? Add/Removal?

Reply:
Hm, there's one attach/detach Info message, which is kept same as with tcx mechanism. I don't have a strong opinion, maybe it could be moved to debug for both at some point but I'd think independently of this set. Probably okay to keep as is for now. Cc @rgo3